torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -159,17 +178,67 @@ void ProcessGroupAgent::join() {
   listenerThread_.join();
 }
 
+bool ProcessGroupAgent::checkNoPendingMessage() {
+  const auto worldSize = pg_->getSize();
+  std::vector<int64_t> snapshot;
+  snapshot.reserve(2 * worldSize);
+  auto recvSnapshot = recvCounts_.snapshot();
+  auto sendSnapshot = sendCounts_.snapshot();
+  snapshot.insert(
+      snapshot.end(),
+      std::make_move_iterator(recvSnapshot.begin()),
+      std::make_move_iterator(recvSnapshot.end()));
+  snapshot.insert(
+      snapshot.end(),
+      std::make_move_iterator(sendSnapshot.begin()),
+      std::make_move_iterator(sendSnapshot.end()));
+
+  std::vector<torch::Tensor> inputSnapshot =
+      {torch::from_blob(snapshot.data(), {2, worldSize}, {torch::kInt64})};
+
+  // allgather both send and recv messages in one shot
+  std::vector<std::vector<torch::Tensor>> outputSnapshots(1);
+
+  for (int i = 0; i < worldSize; ++i) {
+    outputSnapshots[0].emplace_back(
+        torch::zeros({2, worldSize}, {torch::kInt64}));
+  }
+
+  pg_->allgather(outputSnapshots, inputSnapshot)->wait();
+
+  // loop through all send/recv pairs to make sure that all sent messages are
+  // processed.
+  const auto& peerCounts = outputSnapshots[0];
+  for (int from = 0; from < worldSize; ++from) {
+    for (int to = 0; to < worldSize; ++to) {
+      // peerCounts[x][0] is recv counts, and peerCounts[x][1] is send counts
+
+      const auto& sentCnt = peerCounts[from][1][to].data_ptr<int64_t>()[0];
+      const auto& recvCnt = peerCounts[to][0][from].data_ptr<int64_t>()[0];
+      // NB: we cannot throw an error when sentCnt < recvCnt here. Because, send
+      // and recv counts on different workers are read in a distributed manner.
+      // It is possible that the sender reads its send count before sending, but
+      // the receive reads its recv count after receiving. Hence, it is valid to
+      // have sendCnt < recv Cnt.
+      if (sentCnt != recvCnt) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
 void ProcessGroupAgent::sync() {
-  // Block until all processes wants to sync. This is necessary before acquiring
-  // the lock below, because other processes might not enter sync() until it
-  // gets some response from this RpcAgent.
-  pg_->barrier()->wait();
-  // Wait until the all send works are done.
-  // NB: There might be additional send works inserted while waiting.
-  threadPool_.waitWorkComplete();
-  // Use another barrier in case different RpcAgent handles different amounts of
-  // workloads.
+  // Block until all processes wants to sync.
   pg_->barrier()->wait();
+  // block until all peers agree that all sent messages have been processed.
+  do {
+    // Finish all send/recv tasks in the thread pool
+    threadPool_.waitWorkComplete();
+    // As there could be nested RPC calls, or response callback could also
+    // trigger more messages to be sent, we need to wait for the thread pool
+    // again.
+  } while (!checkNoPendingMessage());
```

Comment:
The double negation is not super clear here. I think it would be more readable if the method returned the inverse, like `hasPendingMessage()`, so that the condition reads as "continue while there is a pending message", instead of "continue while there is not no pending messages".